stock_picking_auto_create_lot/models/stock_picking.py

```diff
@@ -12,14 +13,19 @@ def _set_auto_lot(self):
         Allows to be called either by button or through code
         """
         pickings = self.filtered(lambda p: p.picking_type_id.auto_create_lot)
-        lines = pickings.mapped("move_line_ids").filtered(
-            lambda x: (
-                not x.lot_id
-                and not x.lot_name
-                and x.product_id.tracking != "none"
-                and x.product_id.auto_create_lot
-            )
-        )
+        domain = [
+            ("picking_id", "in", pickings.ids),
+            ("lot_id", "=", False),
+            ("lot_name", "=", False),
+            ("product_id.tracking", "!=", "none"),
+            ("product_id.auto_create_lot", "=", True),
+        ]
+        # i.e. There is a move line with done quantity in the picking.
+        # This is to prevent unnecessary generation of unused serials in case of
+        # partial receipts of serial-tracked products.
+        if not pickings._check_immediate():
+            domain = expression.AND([domain, [("qty_done", ">", 0)]])
+        lines = self.env["stock.move.line"].search(domain)
         lines.set_lot_auto()
```

Comment:
Further adjusting what I suggested.  We need to consider the case where mutiple receipt pickings are batch validated, involving one with done quantity on a move line and another one without.

```suggestion
        lines = self.env["stock.move.line"]
        for picking in self:
            if not picking.picking_type_id.auto_create_lot:
                continue
            domain = [
                ("picking_id", "=", picking.id),
                ("lot_id", "=", False),
                ("lot_name", "=", False),
                ("product_id.tracking", "!=", "none"),
                ("product_id.auto_create_lot", "=", True),
            ]
            # i.e. There is a move line with done quantity in the picking.
            # This is to prevent unnecessary generation of unused serials in case of
            # partial receipts of serial-tracked products.
            if not picking._check_immediate():
                domain = expression.AND([domain, [("qty_done", ">", 0)]])
            lines += lines.search(domain)
        if lines:
            lines.set_lot_auto()
```